### PR TITLE
Cass 2.1.15

### PR DESCRIPTION
**Pull request by johadalin**
I think the most important thing in this pull request, is to check that any of the options that been removed/changed are not things that we think are important. 
## Main config changes

Biggest apparent changes are in the env.sh and yaml files.
For these, i took the files from cassandra 2.1.15, and applied the changes that we had previously applied to the 2.0.14 files. On top of those, i then made a few further changes.

The specific changes i have made, as opposed to diffs between cassandra versions, are:
In env.sh:
- Addded `JVM_OPTS="$JVM_OPTS -Dcassandra.ignore_rack=true -Dcassandra.ignore_dc=true"`, as the new version is stricter about overwriting these values, which we do on startup.

In cassandra.yaml:
- Uncommented lines setting directory paths, as casssandra has changed the default location. This included:
  - data_file_directories
  - commitlog_directory
  - saved_caches_directory
- Set the template `broadcast_rpc_address` to 127.0.0.1 so it doesn't break on startup, and overwrite this in the plugin to be the node IP/same as `listen_address`
## Other changes:
- Moved the `touch force_cassandra_yaml_refresh` to a preinst file, rather than the postinst, as this allows us to run it only on upgrade. We don't want to run it on fresh installs, as there are two side-effects:
  - If adding a new node to a cluster, Cassandra 2.1.15 doesn't like the multiple restarts and rejoining clusters, and complains that it is already in the cluster it is attempting to join.
  - More importantly, the previous behaviour was that a new node would join the cluster, have time to get data, and then be destructively restarted (removing data) before rejoining again. If you added <replication-factor> number of nodes or more in one go, there is the possibility that they would all join, get given data, and be the sole possessors of that data, and then hit destructive restart, losing those keys completely.

Cassandra have removed the following options from the .yaml file, which i assume is fine, but may be worth checking. We hadn't ever changed them, so unless something stands out, should be fine as it is:
- memtable_flush_queue_size: 4
- in_memory_compaction_limit_in_mb: 64
- multithreaded_compaction: false
- compaction_preheat_key_cache: true
- preheat_kernel_page_cache: false
## Testing:

In kicking the tires, since working through all the install/setup issues, i have run through the following scenarios, and seen no problems:
- Set up a 2.0.14 cluster, provision data into cassandra, upgrade nodes to 2.1.15, check data still present.
- Set up a 2.1.15 cluster, run all methods of bulk provisioning. Use cqlsh to check data in place correctly.

The correct upgrade process has not yet been figured out, so we can't just merge away straight off.

### CODE REVIEW

**Comment by graemerobertson** [comment]
@mirw has expressed a passion for reviewing this - thanks!

**Comment by mirw** [comment]
Regarding your comment about `JVM_OPTS="$JVM_OPTS -Dcassandra.ignore_rack=true -Dcassandra.ignore_dc=true"`, how do we know this is safe?  Is it possible that Cassandra are being strict about this for a good reason?

**Comment by mirw** [comment]
The `touch force_cassandra_yaml_refresh` change sounds like a bug that exists today.  Should we merge it separately, and announce as a bug fix?

**Comment by mirw** [comment]
Regarding "Cassandra have removed the following options from the .yaml file, which i assume is fine, but may be worth checking."... yes, I think it's worth checking.  I'd suggest not merging until you've done so.

**Comment by mirw** [comment]
With your upgrade testing, do you need the Cassandra 2.0.14 software to have any of the patches included in this pull request (e.g. the `force_cassandra_yaml_refresh` change)?  I don't think so, but please check.

File: modules/clearwater-etcd-plugins [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit 171537e59df83bc8a6699a0c5f41b7486831c9e0
+Subproject commit 5ed531860c65217ed13f0fbbdb2683bc1ac3a71d
```
Comment (mirw) [inline modules/clearwater-etcd-plugins]: What's this change?

**Comment by mirw** [comment]
Looks good - thanks for the clear write-up.  A few comments above, and are you expecting to make changes to clearwater-etcd-plugins?  (I haven't seen a pull request.)

File: modules/clearwater-etcd-plugins (hunk 1 above)
Comment (johadalin) [inline modules/clearwater-etcd-plugins]: This is the change to set broadcast_rpc_address correctly in the cassandra plugin: https://github.com/Metaswitch/clearwater-etcd-plugins/tree/cass_2.1.15

**Comment by sebrexmetaswitch** [comment]
Since I'm doing some more testing of Cassandra 2.1.15, I've looked at some of the above comments:

**JVM_OPTS**
These are just safety checks added in 2.1 to prevent changing rack or datacentre on a node with data. We only need to do this when first adding a node to a cluster, at which point we have no data so we think it's safe

**touch force_cassandra_yaml_refresh**
May be an existing bug but since we're looking to merge this this week, no point splitting out the code now.

**options removed from .yaml file**
These all look safe, and I haven't hit any issues in testing. These were removed either because they were no longer relevant (e.g. ```compaction_preheat_key_cache```) or because the using it was problematic (e.g. ```multithreaded_compaction```) . In either case, we didn't change the defaults.

**do we need 2.0.14 to have any fixes?**
Don't believe so. confirmed by upgrade testing.

**Pull request merged**